### PR TITLE
Local

### PR DESCRIPTION
**Pull request by jainpr94**

### CODE REVIEW

File: Model/Model.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        self.schedule = RandomActivation(self)
+        self.running = True
+
+    def init_data(self, agents_info, df_s_win_np, df_s_sum_np, df_d_win_np, df_d_sum_np,
+                  diff_win, diff_sum, adopt_year):
+
+        for i in range(self.num_agents):
+
+            # canton of each agent
+            canton = agents_info.iloc[i]['canton']
```
Comment (marius-ethz) [inline Model/Model.py:20]: I would suggest changing this structure to:
- import the total number of buildings in Switzerland per canton 
- introduce a resize-population parameter which allows to change the resizing of the total population and the population in each canton. For example, this allows for reducing the resizing for testing purposes.

File: Model/Model.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+            pv_potential_s = df_s_sum_np[:, i]
+
+            # electricity demand winter for each agent
+            el_demand_w = df_d_win_np[:, i]
+
+            # electricity demand summer for each agent
+            el_demand_s = df_d_sum_np[:, i]
+
+            # surplus in winter
+            surplus_w = diff_win[:,i]
```
Comment (marius-ethz) [inline Model/Model.py:38]: not a given parameter. Should be calculated by the agent when calculating the NPV

File: Model/Model.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+            # surplus in summer
+            surplus_s = diff_sum[:, i]
+
+            a = Building(i, self, canton, pv_size, pv_potential_w, pv_potential_s, el_demand_w, el_demand_s,
+                         surplus_w,surplus_s, adopt_year)
```
Comment (marius-ethz) [inline Model/Model.py:44]: The following parameters are not necessary in my opinion:
- surplus_w
- surplus_s
- adopt_year

File: ABM_main.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+# # Import the Agent
+# from Agent.Agents import *
+
+
+# Import Wholesale Electricity Prices
+
+from Tools.wep_interpolation import canton_list
+from Tools.wep_interpolation import winter_wep
+from Tools.wep_interpolation import summer_wep
```
Comment (marius-ethz) [inline ABM_main.py:18]: What is the reasoning for not including the code from wep.interpolation directly in the main?

File: Agent/Agents.py [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+    elec_bill_savings = (sum(hourly_summer_savings) + sum(hourly_winter_savings)) * (365 / 2)
+
+    # Calculate Revenues due to the feed-in-tariff
+    feed_in_tariff = Data_dict['PV_FIT'][agent.canton][0]
+    revenues_feed_in_tariff = (sum(agent.surplus_s) + sum(agent.surplus_w)) * 365 / 2 * feed_in_tariff
+
+    annual_cash_flow = revenues_feed_in_tariff + elec_bill_savings - costs_OM
+
+    annual_cash_flows = np.repeat(annual_cash_flow, tdata_dict["pv_lifetime"])
+    inv_costs = pv_inv_cost_net
```
Comment (marius-ethz) [inline Agent/Agents.py:189]: why?

**Review by marius-ethz — COMMENTED** [review]
Thanks for all your work! 

I have two more general comments in addition to the detailed ones in the code. 

1) Weighted average hours instead of winter and summer lists: I am not sure yet but I remember that there is an alternative approach of having an average winter and an average summer list by heaving weighted solar production hours. I think DistIv is using this to avoid having these two lists. Let's discuss. 

2) It might be faster to update in the beginning of the year parameters to the new year instead of passing long lists that include all years to all agents. Let's discuss. 

Best,
Marius

File: ABM_main.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
+# Standard imports
+import pandas as pd
+import numpy as np
+import json
+import time
+start_time1 = time.time()
```
[the remark this answers is not in the document]
Comment (jainpr94) [reply ABM_main.py:6]: This was just for personal reference to see which functions are taking most long, will be removing it

File: ABM_main.py [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+# Import the Model
+from Model.Model import *
+
+# # Import the Agent
+# from Agent.Agents import *
+
+
+# Import Wholesale Electricity Prices
+
+from Tools.wep_interpolation import canton_list
```
[the remark this answers is not in the document]
Comment (jainpr94) [reply ABM_main.py:16]: I will keep this cantons list and will make sure that the order is replicated if ever in future another cantons list is required. will convert the list as a csv input

File: ABM_main.py (hunk 4 above)
Comment (jainpr94) [reply ABM_main.py:18]: Included in the main

File: ABM_main.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+PV_omcost1 = pd.read_excel('/home/pranjal/PycharmProjects/Master_thesis_ABM/data/Agents_OMcosts.xlsx')
+
+# converting to numpy and dictionaries for faster processing
+
+Grid_tariff = Grid_tariff1.set_index('canton').T.to_dict('list')
+PV_FIT = PV_FIT1.set_index('canton').T.to_dict('list')
+PV_margin = PV_margin1.set_index('canton').T.to_dict('list')
+PV_invcost = PV_invcost1.set_index('Category').T.to_dict('list')
+PV_omcost = PV_omcost1.set_index('Category').T.to_dict('list')
+PV_subsidy = PV_subsidy1.set_index('Category').T.to_dict('list')
```
[the remark this answers is not in the document]
Comment (jainpr94) [reply ABM_main.py:51]: Done

File: ABM_main.py [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+years_list = PV_invcost1.columns.to_list()
+
+# one master dictionary to contain all dictionaries
```
[the remark this answers is not in the document]
Comment (jainpr94) [reply ABM_main.py:54]: As for faster processing, I converted all data into dictionaries, this years_list gives the names of columns in all dataframes for returning the appropriate index. similar concept to cantons_list

File: Agent/Agents.py (hunk 5 above)
Comment (jainpr94) [reply Agent/Agents.py:189]: Corrected

File: ABM_main.py [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+df_s_win_np= np.zeros(shape=(24, 2196))
+z_sum = pd.read_excel('/home/pranjal/PycharmProjects/Master_thesis_ABM/data/PV_total_summer.xlsx')
+z_win = pd.read_excel('/home/pranjal/PycharmProjects/Master_thesis_ABM/data/PV_total_winter.xlsx')
+
+# solar potentials per hour for an average summer day and a winter day (24 hrs) are associated with every agent
+# and stored in different dataframes, one for winter and on for summer
+
+for Filelist in agents_list:
+    df_s_sum_np[:, Filelist] = z_sum.E_PV_gen_kWh
+    df_s_win_np[:, Filelist] = z_win.E_PV_gen_kWh
```
[the remark this answers is not in the document]
Comment (jainpr94) [reply ABM_main.py:80]: This code is assign the 24 hour solar potential and demand to every agent, will be changing it to per canton now

File: ABM_main.py [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+# print(df_solar_win)
+
+df_d_sum_np = np.zeros(shape=(24, 2196))
+df_d_win_np = np.zeros(shape=(24, 2196))
+zz_sum = pd.read_excel('/home/pranjal/PycharmProjects/ABM/Untitled_Message/hourly_demand_summer.xlsx')
+zz_win = pd.read_excel('/home/pranjal/PycharmProjects/ABM/Untitled_Message/hourly_demand_winter.xlsx')
+for Filelist in agents_list:
+    df_d_sum_np[:, Filelist] = zz_sum.GRID_kWh
+    df_d_win_np[:, Filelist] = zz_win.GRID_kWh
+print(time.time() - start_time1)
```
[the remark this answers is not in the document]
Comment (jainpr94) [reply ABM_main.py:91]: Similar use as above, and the last line of the code was to see the time taken to generate theses arrays which I removed now.

File: ABM_main.py [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+
+diff_sum = df_s_sum_np - df_d_sum_np
+diff_win = df_s_win_np - df_d_win_np
+diff_sum[diff_sum <= 0] = 0
+diff_win[diff_win <= 0] = 0
```
[the remark this answers is not in the document]
Comment (jainpr94) [reply ABM_main.py:96]: this is to calculate the surplus PV generated for every agent, now a part of NPV calculation.

File: Agent/Agents.py [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+        self.intention = False
+        self.pv_size = pv_size
+        self.pv_potential_w = pv_potential_w
+        self.pv_potential_s = pv_potential_s
+        self.el_demand_w = el_demand_w
+        self.el_demand_s = el_demand_s
+        self.pv_sc_w = []
+        self.pv_sc_s = []
+        self.surplus_w = surplus_w
+        self.surplus_s = surplus_s
```
[the remark this answers is not in the document]
Comment (jainpr94) [reply Agent/Agents.py:27]: done

File: Agent/Agents.py [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+    # Step
+    def step(self):
+        # Update my parameters
+        self.retail_elec_price_s = np.add(cal_hourly_wholesale_elec_price_s(self, self.adopt_yr),
+                                          other_fees(self)).tolist()
+        self.retail_elec_price_w = np.add(cal_hourly_wholesale_elec_price_w(self, self.adopt_yr),
+                                          other_fees(self)).tolist()
+
+        self.pv_sc_s = cal_selfconsumption_summer(self)
+        self.pv_sc_w = cal_selfconsumption_winter(self)
```
[the remark this answers is not in the document]
Comment (jainpr94) [reply Agent/Agents.py:54]: done

File: Agent/Agents.py [hunk 15]
```diff
@@ -0,0 +1,195 @@
...
+
+    costs_OM = (sum(agent.pv_potential_s) + sum(agent.pv_potential_w)) * (365 / 2) * om_rate
+
+    ######## Calculate Annual Revenues ###########
+    # Calculate electricity bill savings due to lower grid electricity demand as PV electricity is self-consumed
+    hourly_summer_savings = []
+    hourly_winter_savings = []
+    for i in range(24):
+        # Calculate savings for summer
+        summer_savings = cal_selfconsumption_summer(agent)[i] * agent.retail_elec_price_s[i]
```
[the remark this answers is not in the document]
Comment (jainpr94) [reply Agent/Agents.py:173]: Done, there is a separate function for self consumption calculation which calculates it and then I call the function in the NPV calculation

File: Agent/Agents.py [hunk 16]
```diff
@@ -0,0 +1,195 @@
...
+import numpy as np
+from itertools import chain
+
+
+from ABM_main import summer_wep,winter_wep,tdata_dict,canton_list, years_list,Data_dict
+
+class Building(Agent):
+
+    def __init__(self, unique_id, model, canton, pv_size, pv_potential_w, pv_potential_s, el_demand_w, el_demand_s,
+                 surplus_w, surplus_s, adopt_yr):
```
[the remark this answers is not in the document]
Comment (jainpr94) [reply Agent/Agents.py:12]: Done

File: Model/Model.py (hunk 2 above)
Comment (jainpr94) [reply Model/Model.py:38]: done

File: ABM_main.py [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+    tdata_dict = json.loads(myjson.read())
+
+# importing agents list which contains every agent's unique ID, PV size(randomly assigned for now) and canton
+
+agents_info = pd.read_csv('/home/pranjal/PycharmProjects/Master_thesis_ABM/data/Agents_info.csv')
+# change to csv and add csv to data structure
+
+# to iterate over every agent and assign individual solar potential demand
+
+agents_list = list(agents_info.index)
```
[the remark this answers is not in the document]
Comment (jainpr94) [reply ABM_main.py:33]: Not needed anymore

File: ABM_main.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (jainpr94) [reply ABM_main.py:16]: Also, I wanted to ask I am generating the cantons list in the model itself. Do I still need to convert it into a csv ?

File: ABM_main.py [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+
+
+# Import Wholesale Electricity Prices
+
+from Tools.wep_interpolation import canton_list
+from Tools.wep_interpolation import winter_wep
+from Tools.wep_interpolation import summer_wep
+
+# Import Technology data
+tech_data_path = "/home/pranjal/PycharmProjects/Master_thesis_ABM/data/technology_meta.json"
```
[the remark this answers is not in the document]
Comment (jainpr94) [reply ABM_main.py:21]: done

File: ABM_main.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (marius-ethz) [reply ABM_main.py:16]: okay for me.

File: ABM_main.py [hunk 19]
```diff
@@ -0,0 +1,115 @@
...
+
+with open(tech_data_path) as myjson:
+    tdata_dict = json.loads(myjson.read())
+
+# importing agents list which contains every agent's unique ID, PV size(randomly assigned for now) and canton
+
+agents_info = pd.read_csv('/home/pranjal/PycharmProjects/Master_thesis_ABM/data/Agents_info.csv')
```
[the remark this answers is not in the document]
Comment (jainpr94) [reply ABM_main.py:28]: done

File: Model/Model.py (hunk 1 above)
Comment (jainpr94) [reply Model/Model.py:20]: done

File: ABM_main.py [hunk 20]
```diff
@@ -0,0 +1,115 @@
...
+diff_sum = df_s_sum_np - df_d_sum_np
+diff_win = df_s_win_np - df_d_win_np
+diff_sum[diff_sum <= 0] = 0
+diff_win[diff_win <= 0] = 0
+
+
+
+if __name__ == "__main__":
+
+    diffmodel = DiffModel(2196)
```
[the remark this answers is not in the document]
Comment (jainpr94) [reply ABM_main.py:102]: done

File: ABM_main.py [hunk 21]
```diff
@@ -0,0 +1,115 @@
...
+    for i in range(31):
+        print(i)
+        diffmodel.init_data(agents_info=agents_info, df_s_win_np =df_s_win_np, df_s_sum_np = df_s_sum_np,
+                            df_d_win_np = df_d_win_np, df_d_sum_np =df_d_sum_np,
+                            diff_win=diff_win, diff_sum=diff_sum, adopt_year= i)
```
[the remark this answers is not in the document]
Comment (jainpr94) [reply ABM_main.py:107]: removed.

File: Model/Model.py (hunk 3 above)
Comment (jainpr94) [reply Model/Model.py:44]: done

File: ABM_main.py [hunk 22]
```diff
@@ -0,0 +1,115 @@
...
+# importing agents solar potentials and demands
+#   current assumption : that every agent has same demand and same solar potential
+#   one average day for winter i.e October - March and one for summer i.e. April - September ,hence two arrays are
+#   intialised for both solar and demand
+
+
+df_s_sum_np = np.zeros(shape=(24, 2196))
+df_s_win_np= np.zeros(shape=(24, 2196))
+z_sum = pd.read_excel('/home/pranjal/PycharmProjects/Master_thesis_ABM/data/PV_total_summer.xlsx')
+z_win = pd.read_excel('/home/pranjal/PycharmProjects/Master_thesis_ABM/data/PV_total_winter.xlsx')
```
[the remark this answers is not in the document]
Comment (jainpr94) [reply ABM_main.py:73]: done

File: ABM_main.py [hunk 23]
```diff
@@ -0,0 +1,115 @@
...
+agents_list = list(agents_info.index)
+
+# importing excel files as Pandas data frame
+
+PV_invcost1 = pd.read_excel('/home/pranjal/PycharmProjects/Master_thesis_ABM/data/Agents_Invcosts.xlsx')
+PV_subsidy1 = pd.read_excel('/home/pranjal/PycharmProjects/Master_thesis_ABM/data/Agents_subsidies.xlsx')
+PV_FIT1 = pd.read_excel('/home/pranjal/PycharmProjects/Master_thesis_ABM/data/Agents_FeedIT.xlsx')
+Grid_tariff1 = pd.read_excel('/home/pranjal/PycharmProjects/Master_thesis_ABM/data/Agents_grid_test.xlsx')
+PV_margin1 = pd.read_excel('/home/pranjal/PycharmProjects/Master_thesis_ABM/data/Agents_WEPtoretailmargin.xlsx')
+PV_omcost1 = pd.read_excel('/home/pranjal/PycharmProjects/Master_thesis_ABM/data/Agents_OMcosts.xlsx')
```
[the remark this answers is not in the document]
Comment (jainpr94) [reply ABM_main.py:42]: done

**Pull request closed without merging**